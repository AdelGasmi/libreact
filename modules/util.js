import { createElement as h } from 'react';
export const noop = () => { };
export const idx = (obj, accessor) => {
    try {
        return accessor(obj);
    }
    catch (error) {
        return undefined;
    }
};
export const isClient = typeof window === 'object';
export const debounce = (fn, delay) => {
    let timeout;
    return (...args) => {
        clearTimeout(timeout);
        timeout = setTimeout(() => fn(...args), delay);
    };
};
export const on = (obj, ...args) => obj.addEventListener(...args);
export const off = (obj, ...args) => obj.removeEventListener(...args);
export const ns = (name) => `@@libreact/${name}`;
const hasSymbols = typeof Symbol !== 'undefined';
export const sym = (name) => {
    return hasSymbols ? Symbol(name) : ns(name);
};
export const isFn = (fn) => typeof fn === 'function';
export { h };
