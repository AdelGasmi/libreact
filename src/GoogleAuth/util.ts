export type GoogleAPI = any;

let gapiCache: GoogleAPI;

export const getGapi = async (): Promise<GoogleAPI> => {
    if (gapiCache) {
        return gapiCache;
    }

    await new Promise(resolve => {
        const gapicallback = `__gapicb${Date.now().toString(36)}`;
        (window as any)[gapicallback] = () => {
            delete (window as any)[gapicallback];
            gapiCache = (window as any).gapi;
            resolve();
        };

        const script = document.createElement('script');

        script.src = 'https://apis.google.com/js/platform.js?onload=' + gapicallback;
        document.head.appendChild(script);
    });

    return gapiCache;
};

let gapiAuth2Cache;

export const getGapiAuth2 = async (): Promise<any> => {
    if (gapiAuth2Cache) {
        return gapiAuth2Cache;
    }

    const gapi = await getGapi();

    await new Promise(resolve => {
        gapi.load('auth2', () => {
            gapiAuth2Cache = gapi.auth2;
            resolve();
        });
    });

    return gapiAuth2Cache;
};

let GoogleAuthCache;

export const getGapiAuthInstance = async (client_id: string) => {
    if (GoogleAuthCache) {
        return GoogleAuthCache;
    }

    const gapiAuth2 = await getGapiAuth2();

    GoogleAuthCache = await gapiAuth2.init({client_id});

    return GoogleAuthCache;
};
