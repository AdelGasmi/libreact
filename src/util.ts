export const noop = () => {};

export const idx = (obj, accessor) => {
  try {
    return accessor(obj);
  } catch (error) {
    return undefined;
  }
};

export const isClient = typeof window === 'object';
