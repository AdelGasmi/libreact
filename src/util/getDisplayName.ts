const getDisplayName = (Comp: any) => {
  const tipof = typeof Comp;

  switch (tipof) {
    case 'string':
      return Comp;
    case 'function':
      return Comp.displayName || Comp.name || 'Unknown';
    case 'object':
        return `<${Comp.type.displayName || Comp.type.name || String(Comp)}>`;
    default:
      return 'Unknown';
  }
}

export default getDisplayName;
