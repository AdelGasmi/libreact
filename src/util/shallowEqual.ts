const keyList = Object.keys;
const hasProp = Object.prototype.hasOwnProperty;

const shallowEqual: (a, b) => boolean = (a, b) => {
  if (a === b) return true;
  if (!a || !b) return false;
  if (!(a instanceof Object) || !(b instanceof Object)) return false;

  const keys = keyList(a);
  const length = keys.length;

  for (let i = 0; i < length; i++)
    if (!(keys[i] in b)) return false;

  for (let i = 0; i < length; i++)
    if (a[keys[i]] !== b[keys[i]]) return false;

  return length === keyList(b).length;
}

export default