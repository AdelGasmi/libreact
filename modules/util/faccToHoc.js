import { h } from '../util';
import addClassDecoratorSupport from './addClassDecoratorSupport';
const noWrap = (Comp, propName, props, state) => h(Comp, propName ?
    { [propName]: state, ...props } :
    { ...state, ...props });
export const divWrapper = (Comp, propName, props, state) => h('div', null, noWrap(Comp, propName, props, state));
const faccToHoc = (Facc, prop, wrapper = noWrap) => {
    const hoc = (Comp, propName = prop, faccProps = null) => {
        const isClassDecoratorMethodCall = typeof Comp === 'string';
        if (isClassDecoratorMethodCall) {
            return (Klass) => hoc(Klass, Comp, propName);
        }
        const Enhanced = (props) => h(Facc, faccProps, (state) => wrapper(Comp, propName, props, state));
        return addClassDecoratorSupport(Enhanced);
    };
    return hoc;
};
export default faccToHoc;
