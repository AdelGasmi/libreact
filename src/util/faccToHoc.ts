import {h} from '../util';
import addClassDecoratorSupport from './addClassDecoratorSupport';

const faccToHoc = (Facc, prop?: string) => {
  const hoc = (Comp, propName: any = prop, faccProps: object = null) => {
    const isClassDecoratorMethodCall = typeof Comp === 'string';

    if (isClassDecoratorMethodCall) {
      return (Klass) => hoc(Klass, Comp as any, propName as any);
    }

    const Enhanced = (props) =>
      h(Facc, faccProps, (state) =>
        h(Comp,
          propName ?
            {[propName]: state, ...props} :
            {...state, ...props}
        )
      );

    return addClassDecoratorSupport(Enhanced);
  };

  return hoc;
}

export default faccToHoc;
