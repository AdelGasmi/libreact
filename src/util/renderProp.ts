import {isFn} from '../util';

const renderProp = (props, ...args) => {
  if (process.env.NODE_ENV !== 'production') {
    if (typeof props !== 'object') {
      throw new TypeError('renderProp(props, data) first argument must be a prop object.');
    }

    const {children, render} = props;

    if (isFn(children) && isFn(render)) {
      console.warn(
        'Both "render" and "children" are specified for a render-prop component. ' +
        'Children will be used.'
      );
      console.trace();
    }
  }

  const {children, render} = props;

  return isFn(children) ?
    children(...args) :
    isFn(render) ?
      render(...args) :
      (children || null);
};

export default renderProp;
