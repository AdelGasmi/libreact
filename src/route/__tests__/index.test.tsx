import {mount} from 'enzyme';
import {h} from '../../util';
import {Router, Route} from '..';

describe('route', () => {
  describe('<Router>', () => {
    it('is a component', () => {
      expect(Router).toBeInstanceOf(Function);
    });
  });

  describe('<Route>', () => {
    it('is a component', () => {
      expect(Route).toBeInstanceOf(Function);
    });

    it('does basic routing', () => {
      const wrapper = mount(
        <Router route='/foo'>
          <div>
            <Route match='/foo'>
              /foo
            </Route>
            <Route match='/bar'>
              /foo
            </Route>
          </div>
        </Router>
      );

      expect(wrapper.html()).toMatchSnapshot();
    });
  });
});
