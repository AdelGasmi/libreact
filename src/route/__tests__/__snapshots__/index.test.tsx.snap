// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`route <Route> does basic routing 1`] = `"<div><div>/foo</div><!-- react-empty: 3 --></div>"`;
